test/common/http/http2/codec_impl_test.cc

```diff
@@ -947,30 +954,68 @@ TEST_P(Http2CodecImplTest, ConnectionKeepaliveJitter) {
   client_http2_options_.mutable_connection_keepalive()->mutable_timeout()->set_seconds(1);
   client_http2_options_.mutable_connection_keepalive()->mutable_interval_jitter()->set_value(10);
   /*auto timeout_timer = */ new NiceMock<Event::MockTimer>(&client_connection_.dispatcher_);
-  auto send_timer = new Event::MockTimer(&client_connection_.dispatcher_);
+  auto send_timer = new NiceMock<Event::MockTimer>(&client_connection_.dispatcher_);
 
   constexpr std::chrono::milliseconds min_expected(1000);
   constexpr std::chrono::milliseconds max_expected(1099); // 1000ms + 10%
   std::chrono::milliseconds min_observed(5000);
   std::chrono::milliseconds max_observed(0);
   EXPECT_CALL(*send_timer, enableTimer(_, _))
       .WillRepeatedly(Invoke([&](const std::chrono::milliseconds& ms, const ScopeTrackedObject*) {
+        send_timer->enabled_ = true;
         EXPECT_GE(ms, std::chrono::milliseconds(1000));
         EXPECT_LE(ms, std::chrono::milliseconds(1100));
         max_observed = std::max(max_observed, ms);
         min_observed = std::min(min_observed, ms);
       }));
   initialize();
+  ASSERT_TRUE(send_timer->enabled());
 
+  // Don't allow synchronous dispatch, it results in receiving the ping under
+  // the stack of sending it.
   for (uint64_t i = 0; i < 250; i++) {
```

Comment:
Just so I understand, what in this PR made this change required?

Reply:
Oh yeah, this was a really fun one to debug.
So normal code, you send a ping, do a network flush, eventually get a ping. 

our test code, you send a ping and under the stack due to the way the dispatch is set up it synchronously sends to the peer, which synchronously reads, synchronously responds, which is synchronously dispatched, so you receive the ping under the stack of sending the ping, and so do the "received a ping" logic before finishing up the "sending a ping" logic.

Actual prod code won't do a read under the stack of a write, so I just had to tweak tests to be more like prod.